### DISCUSSION TITLE
Dynamically imported 3rd-party modules are included in the vendors chunk

### DISCUSSION THREAD

**Discussion by azangru** [webpack/webpack#16508] [Q&A] [answered]
There's a common optimization technique that suggests combining all third-party libraries into a single file (which people commonly call "vendors"), like so:

```js
optimization: {
  splitChunks: {
    cacheGroups: {
      // commonly shared code i.e. vendor code to be bundled as vendor.js
      commons: {
        test: /node_modules/,
        name: 'vendors',
        chunks: 'all'
      }
    }
  }
}
```

Only now, by checking my site on pagespeedinsights, I realised that this meant that even those libraries that I imported via dynamic imports, hoping that webpack would code-split them into independent files, got included into the vendors file.

Which makes me wonder:
- Is the advice to combine third-party libraries into a single file (cache group) still relevant these days?
- If yes, is there an easy way to tell webpack to only do this for statically imported dependencies, and code-split dynamically imported modules?

**Comment by alexander-akait** [comment, marked as answer]
> Is the advice to combine third-party libraries into a single file (cache group) still relevant these days?

Based on your application size and architecture, we apply a general method that works in most cases, but it all depends on your architecture, there is no right answer here

> If yes, is there an easy way to tell webpack to only do this for statically imported dependencies, and code-split dynamically imported modules?

Why you use `chunks: 'all'`?

**Reply by azangru** [reply]
> Why you use chunks: 'all'?

That was probably the result of copy-pasting from some example without giving it too much thought 😞 What do you think will better suit the case that I described? I checked with the [docs](https://webpack.js.org/plugins/split-chunks-plugin/#optimizationsplitchunks), but the relevant section just lists the available options, and says that "all" can be "particularly powerful" 😄 I can see why someone would pick "all" after reading this endorsement.

In my case, I suppose it would be reasonable to group statically imported modules into a single chunk, while allowing webpack to split out individual chunks for dynamically loaded modules; but it isn't obvious from the docs how to achieve this.

**Reply by alexander-akait** [reply]
> I suppose it would be reasonable to group statically imported modules into a single chunk

Do you mean `initial`?

I can give advices if you provide more infromation (how you expected splitting - text/screenshot/structure), anyway you can try some experiments and look at results and ask me about options and how they work

**Reply by azangru** [reply]
> Do you mean initial?

Yes, thank you. `initial` seems to be just what I was looking for.
